k8s-install/scripts/install_cni_test.go

```diff
@@ -0,0 +1,208 @@
+package scripts_test
+
+import (
+	. "github.com/onsi/ginkgo"
+	. "github.com/onsi/gomega"
+	"github.com/projectcalico/felix/fv/containers"
+
+	"fmt"
+	"io/ioutil"
+	"os"
+	"time"
+)
+
+var secretFile = "tmp/serviceaccount/token"
+
+// runCniContainer will run install-cni.sh.
+// TODO: We should be returning an error if the command fails, there currently is
+// not a way to get that from the container package.
+func runCniContainer(extraArgs ...string) error {
+	// Get the CWD for mounting directories into container.
+	cwd, err := os.Getwd()
+	if err != nil {
+		Fail("Could not get CWD")
+	}
+
+	// Assemble our arguments.
+	args := []string{
+		"--privileged",
+		"-e", "SLEEP=false",
+		"-e", "KUBERNETES_SERVICE_HOST=127.0.0.1",
+		"-e", "KUBERNETES_SERVICE_PORT=8080",
+		"-v", cwd + "/tmp/bin:/host/opt/cni/bin",
+		"-v", cwd + "/tmp/net.d:/host/etc/cni/net.d",
+		"-v", cwd + "/tmp/serviceaccount:/var/run/secrets/kubernetes.io/serviceaccount",
+	}
+	args = append(args, extraArgs...)
+	args = append(args, "calico/cni:latest", "/install-cni.sh")
+
+	c := containers.Run("cni", args...)
+	for !c.Stopped() {
+		time.Sleep(100 * time.Millisecond)
+	}
+
+	return nil
+}
+
+// cleanup uses the calico/cni container to cleanup after itself as it creates
+// things as root.
+func cleanup() {
+	cwd, err := os.Getwd()
+	if err != nil {
+		Fail("Could not get CWD")
+	}
+	containers.Run("cni",
+		"--privileged",
+		"-e", "SLEEP=false",
+		"-e", "KUBERNETES_SERVICE_HOST=127.0.0.1",
+		"-e", "KUBERNETES_SERVICE_PORT=8080",
+		"-v", cwd+"/tmp/bin:/host/opt/cni/bin",
+		"-v", cwd+"/tmp/net.d:/host/etc/cni/net.d",
+		"-v", cwd+"/tmp/serviceaccount:/var/run/secrets/kubernetes.io/serviceaccount",
+		"calico/cni:latest",
+		"sh", "-c", "rm -f /host/opt/cni/bin/* /host/etc/cni/net.d/*")
+}
+
+var _ = BeforeSuite(func() {
+	// Make the directories we'll need for storing files.
+	err := os.MkdirAll("tmp/bin", 0755)
+	if err != nil {
+		Fail("Failed to create directory tmp/bin")
+	}
+	err = os.MkdirAll("tmp/net.d", 0755)
+	if err != nil {
+		Fail("Failed to create directory tmp/net.d")
+	}
+	err = os.MkdirAll("tmp/serviceaccount", 0755)
+	if err != nil {
+		Fail("Failed to create directory tmp/serviceaccount")
+	}
+	cleanup()
+
+	// Create a secrets file for parsing.
+	k8sSecret := []byte("my-secret-key")
+	err = ioutil.WriteFile(secretFile, k8sSecret, 0755)
+	if err != nil {
+		Fail(fmt.Sprintf("Failed to write k8s secret file: %v", err))
+	}
+})
+
+var _ = AfterSuite(func() {
+	err := os.RemoveAll("tmp")
+	if err != nil {
+		fmt.Println("Failed to clean up directories")
+	}
+})
+
+var _ = Describe("install-cni.sh tests", func() {
+	AfterEach(func() {
+		cleanup()
+	})
+
+	Describe("Run install-cni", func() {
```

Comment:
I think we should have one test case which expects the install-cni.sh command to fail.

Let's run a test where we set:
```
KUBERNETES_SERVICE_HOST=""
KUBERENTES_SERVICE_PORT=""
```

We've hit bugs in the past where those values aren't set and we added special code so that we fail in that case.

https://github.com/projectcalico/cni-plugin/blob/master/k8s-install/scripts/install-cni.sh#L117

Reply:
I can add the test, but at the moment there is no way to get the error codes back from the container, so I had to drop a couple of the tests which check for it.

Reply:
there's no way to tell if a container failed or succeeded?

hmm...

Reply:
So, I was trying the `c.DockerInspect("{{.State.ExitCode}}")` command, but it was after verifying the container was closed, which is after the container has been destroyed and errors out with `Error: No such object`. If you run the inspect before it is closed then it grabs the exit status before the script has finished, so it always returns 0, I verified this by forcing the script to `exit 1` which is reported by the logger, but the docker inspect shows 0:

```shell
# docker inspect output with no format
[
    {
        ...
        "State": {
            ...
            "ExitCode": 0,
            ...
            }
      }
]

# log message
INFO[0013] Container stopped                             error="exit status 1" name=cni-838-6-felixfv
```

I might be missing something obvious, but AFAICT there isn't a way to check it using the current API for the Container.